File: 7.lifecycle/src/lifecycle_cpp_pkg/src/sensor_station.cpp
#include <chrono>
#include <memory>
#include <string>

#include "lifecycle_msgs/msg/state.hpp"
#include "lifecycle_msgs/msg/transition.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_lifecycle/lifecycle_node.hpp"
#include "rclcpp_lifecycle/lifecycle_publisher.hpp"
#include "std_msgs/msg/string.hpp"

using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

class SensorStation : public rclcpp_lifecycle::LifecycleNode
{
public:
  explicit SensorStation(const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
  : rclcpp_lifecycle::LifecycleNode("sensor_station", options), count_(0)
  {
    // Declare parameters in constructor
    this->declare_parameter<std::string>("sensor_name", "sensor_station");
    this->declare_parameter<std::string>("topic_name", "sensor_data");
    this->declare_parameter<double>("publish_frequency", 1.0);

    RCLCPP_INFO(
      get_logger(),
      "[%s] Constructed in UNCONFIGURED state.",
      get_name());
  }

  // 1. Transition: Unconfigured -> Configuring -> Inactive
  CallbackReturn on_configure(const rclcpp_lifecycle::State & /*state*/) override
  {
    sensor_name_ = this->get_parameter("sensor_name").as_string();
    topic_name_ = this->get_parameter("topic_name").as_string();
    publish_frequency_ = this->get_parameter("publish_frequency").as_double();

    if (publish_frequency_ <= 0.0) {
      RCLCPP_ERROR(get_logger(), "[%s] Publish frequency must be > 0.0!", get_name());
      return CallbackReturn::FAILURE;
    }

    // Create Lifecycle Publisher (starts in inactive state by default)
    pub_ = this->create_publisher<std_msgs::msg::String>(topic_name_, 10);

    // Create periodic timer
    auto timer_period = std::chrono::duration<double>(1.0 / publish_frequency_);
    timer_ = this->create_wall_timer(
      timer_period,
      std::bind(&SensorStation::timer_callback, this));

    RCLCPP_INFO(
      get_logger(),
      "[%s] on_configure: Initialized publisher on topic '%s' at %.1f Hz. State is now INACTIVE.",
      get_name(), topic_name_.c_str(), publish_frequency_);

    return CallbackReturn::SUCCESS;
  }

  // 2. Transition: Inactive -> Activating -> Active
  CallbackReturn on_activate(const rclcpp_lifecycle::State & /*state*/) override
  {
    // Explicitly activate the lifecycle publisher so it starts transmitting
    if (pub_) {
      pub_->on_activate();
    }

    RCLCPP_INFO(
      get_logger(),
      "[%s] on_activate: LifecyclePublisher is now ACTIVE and broadcasting messages.",
      get_name());

    return CallbackReturn::SUCCESS;
  }

  // 3. Transition: Active -> Deactivating -> Inactive
  CallbackReturn on_deactivate(const rclcpp_lifecycle::State & /*state*/) override
  {
    // Explicitly deactivate lifecycle publisher so it stops transmitting
    if (pub_) {
      pub_->on_deactivate();
    }

    RCLCPP_INFO(
      get_logger(),
      "[%s] on_deactivate: LifecyclePublisher is now INACTIVE (messages suppressed).",
      get_name());

    return CallbackReturn::SUCCESS;
  }

  // 4. Transition: Inactive -> CleaningUp -> Unconfigured
  CallbackReturn on_cleanup(const rclcpp_lifecycle::State & /*state*/) override
  {
    timer_.reset();
    pub_.reset();
    count_ = 0;

    RCLCPP_INFO(
      get_logger(),
      "[%s] on_cleanup: Destroyed publisher and timer. State is now UNCONFIGURED.",
      get_name());

    return CallbackReturn::SUCCESS;
  }

  // 5. Transition: Any state -> ShuttingDown -> Finalized
  CallbackReturn on_shutdown(const rclcpp_lifecycle::State & /*state*/) override
  {
    timer_.reset();
    pub_.reset();

    RCLCPP_INFO(
      get_logger(),
      "[%s] on_shutdown: Resources cleaned up. State is now FINALIZED.",
      get_name());

    return CallbackReturn::SUCCESS;
  }

  // 6. Transition: Error occurred -> ErrorProcessing
  CallbackReturn on_error(const rclcpp_lifecycle::State & /*state*/) override
  {
    RCLCPP_ERROR(get_logger(), "[%s] on_error: Error detected, recovering...", get_name());
    timer_.reset();
    pub_.reset();
    return CallbackReturn::SUCCESS;
  }

private:
  void timer_callback()
  {
    std_msgs::msg::String msg;
    msg.data = "[" + sensor_name_ + "] reading #" + std::to_string(count_++);

    // LifecyclePublisher::is_activated() returns true only when in ACTIVE state
    if (pub_ && pub_->is_activated()) {
      RCLCPP_INFO(get_logger(), "[ACTIVE] Publishing -> '%s'", msg.data.c_str());
      pub_->publish(msg);
    } else {
      RCLCPP_INFO_THROTTLE(
        get_logger(),
        *get_clock(),
        2000,
        "[%s - INACTIVE] Timer firing, but LifecyclePublisher is disabled (messages suppressed)",
        get_name());
    }
  }

  std::string sensor_name_;
  std::string topic_name_;
  double publish_frequency_{1.0};
  uint64_t count_{0};

  std::shared_ptr<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::String>> pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SensorStation>();
  rclcpp::spin(node->get_node_base_interface());
  rclcpp::shutdown();
  return 0;
}

File: 7.lifecycle/src/lifecycle_py_pkg/lifecycle_py_pkg/sensor_station.py
#!/usr/bin/env python3
import rclpy
from rclpy.lifecycle import LifecycleNode, State, TransitionCallbackReturn
from std_msgs.msg import String


class SensorStation(LifecycleNode):
    def __init__(self, **kwargs):
        super().__init__("sensor_station", **kwargs)
        self.declare_parameter("sensor_name", "sensor_station")
        self.declare_parameter("topic_name", "sensor_data")
        self.declare_parameter("publish_frequency", 1.0)

        self.pub = None
        self.timer = None
        self.count = 0

        self.get_logger().info(f"[{self.get_name()}] Constructed in UNCONFIGURED state.")

    def on_configure(self, state: State) -> TransitionCallbackReturn:
        self.sensor_name = (
            self.get_parameter("sensor_name").get_parameter_value().string_value
        )
        self.topic_name = (
            self.get_parameter("topic_name").get_parameter_value().string_value
        )
        self.publish_frequency = (
            self.get_parameter("publish_frequency").get_parameter_value().double_value
        )

        if self.publish_frequency <= 0.0:
            self.get_logger().error(f"[{self.get_name()}] Publish frequency must be > 0.0!")
            return TransitionCallbackReturn.FAILURE

        # Create Lifecycle Publisher (inactive by default)
        self.pub = self.create_lifecycle_publisher(String, self.topic_name, 10)

        # Create periodic timer
        timer_period = 1.0 / self.publish_frequency
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.get_logger().info(
            f"[{self.get_name()}] on_configure: Initialized publisher on topic '{self.topic_name}' "
            f"at {self.publish_frequency:.1f} Hz. State is now INACTIVE."
        )
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info(
            f"[{self.get_name()}] on_activate: LifecyclePublisher is now ACTIVE and broadcasting messages."
        )
        # super().on_activate activates registered lifecycle publishers
        return super().on_activate(state)

    def on_deactivate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info(
            f"[{self.get_name()}] on_deactivate: LifecyclePublisher is now INACTIVE (messages suppressed)."
        )
        # super().on_deactivate deactivates registered lifecycle publishers
        return super().on_deactivate(state)

    def on_cleanup(self, state: State) -> TransitionCallbackReturn:
        if self.timer is not None:
            self.destroy_timer(self.timer)
            self.timer = None
        if self.pub is not None:
            self.destroy_publisher(self.pub)
            self.pub = None
        self.count = 0

        self.get_logger().info(
            f"[{self.get_name()}] on_cleanup: Destroyed publisher and timer. State is now UNCONFIGURED."
        )
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state: State) -> TransitionCallbackReturn:
        if self.timer is not None:
            self.destroy_timer(self.timer)
            self.timer = None
        if self.pub is not None:
            self.destroy_publisher(self.pub)
            self.pub = None

        self.get_logger().info(
            f"[{self.get_name()}] on_shutdown: Resources cleaned up. State is now FINALIZED."
        )
        return TransitionCallbackReturn.SUCCESS

    def on_error(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().error(f"[{self.get_name()}] on_error: Error detected, recovering...")
        if self.timer is not None:
            self.destroy_timer(self.timer)
            self.timer = None
        if self.pub is not None:
            self.destroy_publisher(self.pub)
            self.pub = None
        return TransitionCallbackReturn.SUCCESS

    def timer_callback(self):
        msg = String()
        msg.data = f"[{self.sensor_name}] reading #{self.count}"
        self.count += 1

        if self.pub is not None and self.pub.is_activated:
            self.get_logger().info(f"[ACTIVE] Publishing -> '{msg.data}'")
            self.pub.publish(msg)
        else:
            self.get_logger().info(
                f"[{self.get_name()} - INACTIVE] Timer firing, but LifecyclePublisher is disabled (messages suppressed)",
                throttle_duration_sec=2.0,
            )


def main(args=None):
    rclpy.init(args=args)
    node = SensorStation()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
